framework/src/tinyalsa/tinyalsa.c

```diff
@@ -1075,25 +1136,24 @@ int pcm_start(struct pcm *pcm)
 	if (pcm == NULL) {
 		return -EINVAL;
 	}
+
+	if (pcm->running) {
+		return 0;
+	}
+
 	int prepare_error = pcm_prepare(pcm);
 	if (prepare_error) {
 		return prepare_error;
 	}
 
+	/* If the device is opened for read and Our buffers are not enqued. Enque them now. */
 	if (pcm->flags & PCM_IN) {
-		/* If the device is opened for read and Our buffers are not enqued. Enque them now. */
 #ifdef CONFIG_AUDIO_MULTI_SESSION
 		bufdesc.session = pcm->session;
 #endif
-#ifdef CONFIG_AUDIO_DRIVER_SPECIFIC_BUFFERS
 		bufdesc.numbytes = pcm->config.period_size;
-		for (pcm->bufPtr = 0; pcm->bufPtr < buf_info.nbuffers; pcm->bufPtr++)
-#else
-		bufdesc.numbytes = CONFIG_AUDIO_BUFFER_NUMBYTES;
-		for (pcm->bufPtr = 0; pcm->bufPtr < CONFIG_AUDIO_NUM_BUFFERS; pcm->bufPtr++)
-#endif
-		{
-			bufdesc.u.pBuffer = pcm->pBuffers[pcm->bufPtr];
+		while (pcm->buf_ptr < pcm->config.period_count) {
+			bufdesc.u.pBuffer = pcm->pBuffers[pcm->buf_ptr++];
```

Comment:
I would advise to postpone this change . The current tc needs to be fixed in setting correct period parameters in set_config call. This change causes a crash with audio_tc. We need to figure out the way to manage period_size aned period_count .